pkg/consensus/payload.go

```diff
@@ -0,0 +1,153 @@
+package consensus
+
+import (
+	"fmt"
+
+	"github.com/CityOfZion/neo-go/pkg/core/transaction"
+	"github.com/CityOfZion/neo-go/pkg/crypto/hash"
+	"github.com/CityOfZion/neo-go/pkg/io"
+	"github.com/CityOfZion/neo-go/pkg/util"
+	"github.com/pkg/errors"
+)
+
+type (
+	messageType byte
+
+	message struct {
+		Type       messageType
+		ViewNumber byte
+
+		payload io.Serializable
+	}
+
+	Payload struct {
+		message
+
+		Version        uint32
+		ValidatorIndex uint16
+		PrevHash       util.Uint256
+		Height         uint32
+		Timestamp      uint32
+
+		Witness transaction.Witness
+	}
+)
+
+const (
+	changeViewType      messageType = 0x00
+	prepareRequestType  messageType = 0x20
+	prepareResponseType messageType = 0x21
+	commitType          messageType = 0x30
+	recoveryRequestType messageType = 0x40
+	recoveryMessageType messageType = 0x41
+)
+
+func (p *Payload) EncodeBinaryUnsigned(w *io.BinWriter) {
+	w.WriteLE(p.Version)
+	w.WriteBE(p.PrevHash[:])
+	w.WriteLE(p.Height)
+	w.WriteLE(p.ValidatorIndex)
+	w.WriteLE(p.Timestamp)
+
+	ww := io.NewBufBinWriter()
+	p.message.EncodeBinary(ww.BinWriter)
+	w.WriteBytes(ww.Bytes())
+}
+
+func (p *Payload) EncodeBinary(w *io.BinWriter) {
+	p.EncodeBinaryUnsigned(w)
+
+	w.WriteLE(byte(1))
+	p.Witness.EncodeBinary(w)
+}
+
+func (p *Payload) DecodeBinaryUnsigned(r *io.BinReader) {
+	r.ReadLE(&p.Version)
+	r.ReadBE(p.PrevHash[:])
+	r.ReadLE(&p.Height)
+	r.ReadLE(&p.ValidatorIndex)
+	r.ReadLE(&p.Timestamp)
+
+	data := r.ReadBytes()
+	rr := io.NewBinReaderFromBuf(data)
+	p.message.DecodeBinary(rr)
+
+	return
+}
+
+func (p *Payload) Hash() util.Uint256 {
+	data := p.getHashData()
+	return hash.DoubleSha256(data)
+}
+
+func (p *Payload) getHashData() []byte {
+	w := io.NewBufBinWriter()
+	p.EncodeBinaryUnsigned(w.BinWriter)
+
+	return w.Bytes()
+}
+
+func (p *Payload) DecodeBinary(r *io.BinReader) {
+	p.DecodeBinaryUnsigned(r)
+
+	var b byte
+	r.ReadLE(&b)
+	if b != 1 {
+		r.Err = errors.New("invalid format")
+		return
+	}
+
+	p.Witness.DecodeBinary(r)
+}
+
+func (m *message) EncodeBinary(w *io.BinWriter) {
+	w.WriteLE(byte(m.Type))
+	w.WriteLE(m.ViewNumber)
+	m.payload.EncodeBinary(w)
+}
+
+func (m *message) DecodeBinary(r *io.BinReader) {
+	r.ReadLE((*byte)(&m.Type))
+	r.ReadLE(&m.ViewNumber)
+
+	switch m.Type {
+	case changeViewType:
+		cv := new(changeView)
+		// NewViewNumber is not marshaled
```

Comment:
why is that so?

Reply:
Because C# implementation doesn't do it.